src/blog/internet-computer-web-app-decentralized-database-architecture.md

```diff
@@ -0,0 +1,345 @@
+
+If we develop, for example, an application that list the kind of animals our users own, we would most probably define a db collection of `pets` to collect their entries.
+
+![](https://cdn-images-1.medium.com/max/1600/1*I4CP0N8GvViWCuAzRLa3UA.png)
+
+To query and persist the data, we would then implement a getter (`get` ) and setter (`add` ) functions in our web application that would call the cloud database.
+
+```javascript
+import firebase from 'firebase/app';
+import 'firebase/firestore';
+
+const firestore = firebase.firestore();
+
+const get = async (entryId) => {
+  const snapshot = 
+            await firestore.collection('pets').doc(entryId).get();
+
+  if (snapshot.exists) {
+    console.log(snapshot.data());
+  }
+};
+
+const add = async (userId, dog, cat) => {
+  const {id: entryId} = await firestore.collection('pets').add({
+    userId,
+    dog,
+    cat
+  });
+
+  console.log(entryId);
+};
+```
+
+All users would use the same entry point and, their data would also be saved in the same collection (`firestore.collection('pets')` ).
+
+*****
+
+### The “Conservative”
+
+The Internet Computer does not provide a built-in out of the box database solution, that is not the goal of the DFINITY foundation (my understanding). They provide a futuristic blockchain network, that can run web applications too.
+
+It is up to the makers to build features on the top. They can implement features and, unleash these on the IC through canister smart contracts.
+
+I picture my self a "canister" has a container that run a set of functions in the cloud, except that it is actually a decentralized blockchain network 😉. It can basically do anything that can be implemented, from computing to data persistence.
+
+After studying the [documentation](https://sdk.dfinity.org/docs/introduction/welcome.html) of the SDK, [examples](https://github.com/dfinity/examples) and the “[linkedup](https://github.com/dfinity/linkedup/)” demo,  our first approach had then for goal to replicate the approach we are familiar with, the one exposed previously, the “Nowadays".
+
+![](https://cdn-images-1.medium.com/max/1600/1*GK30d2fOgC-7UVWdAnrQFA.png)
+
+However, given the blockchain nature of the Internet Computer, it may be good to imagine the architecture in a decentralized way with data shared in the form of blocks.
+
+![](https://cdn-images-1.medium.com/max/1600/1*jI-CDFohigqIZCibwWrcqw.png)
+
+If we omit the chain aspect and, the decentralization, it basically works the same as what I am familiar with isn't?
+
+*****
+
+#### Code Snippet
+
+The way we can picture the persistence of the data within a canister, we would develop for our "pets" web application, would be again quite similar to previous architecture.
+
+![](https://cdn-images-1.medium.com/max/1600/1*DcChK7ojaMYGqExH6F-YtQ.png)
+
+Instead of key-value database collection, as there are no “built-in out of the box” database available, our actor would contain a variable that would act as such.
```

Comment:
```suggestion
Instead of a key-value database collection -- as there are no built-in databases available -- our actor would simply have an attribute holding the data.
```